geomesa-spring/geomesa-spring-cloud-stream-binder-kafka-datastore/src/main/java/org/locationtech/geomesa/spring/binder/kafka/datastore/KafkaDatastoreMessageBinder.java

```diff
@@ -0,0 +1,111 @@
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.spring.binder.kafka.datastore;
+
+import org.geotools.api.data.DataStore;
+import org.geotools.api.data.FeatureWriter;
+import org.geotools.api.data.SimpleFeatureStore;
+import org.geotools.api.data.Transaction;
+import org.geotools.api.feature.simple.SimpleFeature;
+import org.geotools.api.feature.simple.SimpleFeatureType;
+import org.geotools.api.filter.Filter;
+import org.geotools.api.filter.FilterFactory;
+import org.geotools.factory.CommonFactoryFinder;
+import org.locationtech.geomesa.kafka.utils.KafkaFeatureEvent;
+import org.locationtech.geomesa.utils.geotools.FeatureUtils;
+import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypeLoader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.cloud.stream.binder.AbstractMessageChannelBinder;
+import org.springframework.cloud.stream.binder.ConsumerProperties;
+import org.springframework.cloud.stream.binder.ProducerProperties;
+import org.springframework.cloud.stream.provisioning.ConsumerDestination;
+import org.springframework.cloud.stream.provisioning.ProducerDestination;
+import org.springframework.integration.core.MessageProducer;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.MessageHandler;
+
+import java.io.IOException;
+import java.util.Set;
+import java.util.function.Supplier;
+
+public class KafkaDatastoreMessageBinder extends AbstractMessageChannelBinder<ConsumerProperties, ProducerProperties, KafkaDatastoreBinderProvisioner> {
+    private static final Logger logger = LoggerFactory.getLogger(KafkaDatastoreMessageBinder.class);
+
+    private final Supplier<DataStore> dsFactory;
+    private final DataStore ds;
+    FeatureWriter<SimpleFeatureType, SimpleFeature> writer;
+    FilterFactory ff;
+
+    public KafkaDatastoreMessageBinder(
+            String[] headersToEmbed,
+            KafkaDatastoreBinderProvisioner provisioningProvider,
+            Supplier<DataStore> dsFactory
+    ) {
+        super(headersToEmbed, provisioningProvider);
+        this.dsFactory = dsFactory;
+        this.ds = dsFactory.get();
+        ff = CommonFactoryFinder.getFilterFactory();
+    }
+
+    // Maybe handle a Collection<SimpleFeature> and a Collection<KafkaFeatureEvent>
+    @Override
+    protected MessageHandler createProducerMessageHandler(
+            final ProducerDestination destination,
+            final ProducerProperties producerProperties,
+            final MessageChannel errorChannel) {
+        return message -> {
+            String sfName = destination.getName();
+            SimpleFeature payload;
+
+            try {
+                var sft = SimpleFeatureTypeLoader.sftForName(sfName);
+                if (sft.isDefined()) {
+                    ds.createSchema(sft.get());
```

Comment:
you might want to verify here that the sft matches any existing schema in the store - it's a bit tricky to do that well though :/

Reply:
I thought createSchema created it if it was missing and did nothing if it already existed.

Reply:
right, but you can have issues if the schema exists in the store, but it doesn't match the one loaded by the SimpleFeatureTypeLoader. maybe in this case it doesn't actually matter...